modules/videoio/src/cap_gstreamer.cpp

```diff
@@ -1715,6 +1721,31 @@ bool GStreamerCapture::open(const String &filename_, const cv::VideoCaptureParam
 
             fps = (double)num/(double)denom;
 
+            GstQuery *query = gst_query_new_duration(GST_FORMAT_DEFAULT);
+            gboolean res = gst_element_query(pipeline.get(), query);
+
+            if (res) {
+                gst_query_parse_duration(query, NULL, &duration);
+            } else
+            if (fps != 0) {
+                GstQuery *query2 = gst_query_new_duration(GST_FORMAT_TIME);
+                gboolean res2 = gst_element_query(pipeline.get(), query2);
+
+                if (res2) {
+                    gst_query_parse_duration(query2, NULL, &duration);
+                    duration = static_cast<gint64>((float)duration / 1000000000*fps);
```

Comment:
What aboud VFR videos? How will it work?

Reply:
Now I made it changed and made it clear in the commit message. VFR does not calculate frame count and return 0 for fps. If fps query fails fps equals to -1.